# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/TransceiverImpl.cpp
# Commit: c137938d4226
# Full Hash: c137938d422606ad8ee3bceb50dac639fa2a562d
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Make MediaPipeline natively async with WatchManager. r=bwc
#   
#   On its own this patch should mainly be seen as a refactor of the way we handle
#   races in MediaPipelineTransmit::SetTrack. Namely that Stop() immediately
#   followed by SetTrack was exposed to racing prior to this patch.
# ==============================================================================

diff -r af4c24767e48 -r c137938d4226 dom/media/webrtc/jsapi/TransceiverImpl.cpp
--- a/dom/media/webrtc/jsapi/TransceiverImpl.cpp	Fri Apr 09 10:26:12 2021 +0200
+++ b/dom/media/webrtc/jsapi/TransceiverImpl.cpp	Tue Apr 13 15:02:15 2021 +0200
@@ -264,6 +264,9 @@
 
   mTransmitPipeline->Stop();
 
+  mConduit->StopReceiving();
+  mConduit->StopTransmitting();
+
   // NOTE(pkerr) - the Call API requires the both local_ssrc and remote_ssrc be
   // set to a non-zero value or the CreateVideo...Stream call will fail.
   if (mJsepTransceiver->mSendTrack.GetSsrcs().empty()) {
@@ -301,6 +304,7 @@
   }
 
   if (mJsepTransceiver->mRecvTrack.GetActive()) {
+    mConduit->StartReceiving();
     mReceiver->Start();
   }
 
@@ -310,6 +314,8 @@
                 mPCHandle << "[" << mMid << "]: " << __FUNCTION__
                           << " Starting transmit conduit without send track!");
     }
+
+    mConduit->StartTransmitting();
     mTransmitPipeline->Start();
   }
 
@@ -919,7 +925,7 @@
 }
 
 void TransceiverImpl::Stop() {
-  mTransmitPipeline->Shutdown_m();
+  mTransmitPipeline->Shutdown();
   mReceiver->Shutdown();
   // Make sure that stats queries stop working on this transceiver.
   UpdateSendTrack(nullptr);